include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+                                                        operation)]() mutable {
+                beast::error_code error{};
+                operation(error, 0);
+            });
+        }
+        return init.result.get();
+    }
+
+    template <typename ConstBufferSequence, typename Handler>
+    auto async_write_some(ConstBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        async_write_some_op<ConstBufferSequence, Handler> operation{
+            *this, sequence, std::move(init.completion_handler)};
+        next_layer().get_io_service().dispatch([operation = std::move(
+                                                    operation)]() mutable {
+            beast::error_code error{};
+            operation(error, 0);
+        });
+        return init.result.get();
+    }
+
+    void set_secrets(const crypto::aead_secrets& secrets) {
+        this->secrets = secrets;
+    }
+
+private:
+    template <typename ConstBufferSequence, typename Handler>
+    struct async_write_some_op : public asio::coroutine {
+        async_write_some_op(secure_socket& socket,
+                            const ConstBufferSequence& sequence,
+                            Handler&& handler)
+            : socket{socket},
+              sequence{sequence},
+              handler{std::forward<Handler>(handler)} {}
+
+        void operator()(beast::error_code error, size_t n) {
+            CORO_REENTER(*this) {
+                static_cast<void>(n);
+                socket.encrypt_pdu(sequence, pdu);
+                CORO_YIELD asio::async_write(socket.next_layer(),
+                                             asio::buffer(pdu), *this);
+                handler(error, asio::buffer_size(sequence));
+            }
+        }
+
+        secure_socket& socket;
+        ConstBufferSequence sequence;
+        Handler handler;
+        std::vector<gsl::byte> pdu;
+    };
+
+    template <typename MutableBufferSequence, typename Handler>
+    struct async_read_some_op : public asio::coroutine {
+        async_read_some_op(secure_socket& socket,
+                           const MutableBufferSequence& sequence,
+                           Handler&& handler)
+            : socket{socket},
+              sequence{sequence},
+              handler{std::forward<Handler>(handler)} {}
+        secure_socket& socket;
+        MutableBufferSequence sequence;
+        Handler handler;
+
+        size_t transferred = 0;
+        boost::endian::little_int16_t payload_length = 0;
+        boost::optional<beast::flat_buffer::mutable_buffers_type> raw_sequence;
+
+        void operator()(beast::error_code error, size_t n) {
+            CORO_REENTER(*this) {
+                while (transferred < sizeof(payload_length) + payload_length +
+                                         crypto::aead_auth_tag_size) {
+                    raw_sequence = socket.raw_buffer.prepare(1024);
```

Comment:
As before, wrap this in try/catch to convert `std::length_error const&` to an `error_code`